### PR TITLE
SH CAS Required Attribute

### PR DESCRIPTION
**Pull request by stevenhammerton**
Add optional CAS config for a required attribute (and further optional for required attribute value) to restrict which CAS users can log in to synapse.

Any questions, let me know.

Signed-off-by: Steven Hammerton steven.hammerton@openmarket.com

### CODE REVIEW

File: synapse/config/cas.py [hunk 1]
```diff
@@ -27,13 +27,28 @@ def read_config(self, config):
...
+
+            if "required_attribute" in cas_config:
+                self.cas_required_attribute = cas_config["required_attribute"]
+            else:
+                self.cas_required_attribute = None
+
+            if "required_attribute_value" in cas_config:
+                self.cas_required_attribute_value = cas_config["required_attribute_value"]
+            else:
+                self.cas_required_attribute_value = None
```
Comment (erikjohnston) [inline synapse/config/cas.py]: FYI, you can do:

```python
self.cas_required_attribute_value = cas_config.get("required_attribute_value", None)
```

**Comment by erikjohnston** [comment]
Would it make sense to format the config as...?

```yaml
cas_config:
  server_url: "https://cas-server.com"
  required_atrributes:
    <name>: <value>
    ...
```

File: synapse/rest/client/v1/login.py [hunk 2]
```diff
@@ -125,6 +126,47 @@ def do_password_login(self, login_submission):
 
     @defer.inlineCallbacks
     def do_cas_login(self, cas_response_body):
+        (user, attributes) = self.parse_cas_response(cas_response_body)
```
Comment (erikjohnston) [inline synapse/rest/client/v1/login.py]: The left parenthesis are technically not required here.

**Comment by erikjohnston** [comment]
@matrixbot ok to test

File: synapse/rest/client/v1/login.py [hunk 3]
```diff
@@ -133,33 +175,17 @@ def do_cas_login(self, cas_response_body):
...
-                    }
-
-                defer.returnValue((200, result))
+            if child.tag.endswith("attributes"):
+                attributes = {}
+                for attribute in child:
+                    if "}" in attribute.tag:
+                        attributes[attribute.tag.split("}")[1]] = attribute.text
+                    else:
+                        attributes[attribute.tag] = attribute.text
```
Comment (erikjohnston) [inline synapse/rest/client/v1/login.py]: What's going on here with the `}`?

Comment (stevenhammerton) [inline synapse/rest/client/v1/login.py]: The ET XML library expand namespaces to {namespace-url} in element tags. (https://docs.python.org/2/library/xml.etree.elementtree.html#parsing-xml-with-namespaces)

I'm ignoring namespaces here.

File: synapse/config/cas.py (hunk 1 above)
Comment (stevenhammerton) [inline synapse/config/cas.py]: Neat, I'll change.

Comment (stevenhammerton) [inline synapse/config/cas.py]: Fixed in 01a5f19

**Comment by stevenhammerton** [comment]
Agree on the different config format, changed in 01a5f19

File: synapse/rest/client/v1/login.py (hunk 2 above)
Comment (stevenhammerton) [inline synapse/rest/client/v1/login.py]: Fixed in 7f8fdc9

**Comment by erikjohnston** [comment]
Oops, meant to comment on the PR rather than the commit.

**Comment by stevenhammerton** [comment]
Nice, can you tell I don't really know python but really want to use Matrix with CAS :)

83b464e

File: synapse/rest/client/v1/login.py (hunk 3 above)
Comment (erikjohnston) [inline synapse/rest/client/v1/login.py]: Could you add this as a quick comment please?

**Comment by erikjohnston** [comment]
> Nice, can you tell I don't really know python but really want to use Matrix with CAS :)

Looks like you're learning it fast though :D 


Other than adding a quick comment about the namespacing thing, this LGTM :)

**Comment by stevenhammerton** [comment]
Comment added in 739464f

**Comment by erikjohnston** [comment]
Yay! LGTM!

**Comment by NegativeMjark** [comment]
LGTM

**Comment by erikjohnston** [comment]
(FYI: You can also do `return foo, bar`, but @NegativeMjark wasn't convinced that I should teach you this)

**Pull request merged**